site/content/blog/2019-10-21-localizing-matterpoll/index.md

```diff
@@ -0,0 +1,182 @@
+
+- Better support for plural forms with options for e.g. `Zero`, `One`, `Few` and `Many`
+- No global state, which helps with running tests in parallel
+- Usage of [`golang.org/x/text/language`](https://godoc.org/golang.org/x/text/language) for standardized behavior
+
+For further information please take a look at the [Changelog](https://github.com/nicksnyder/go-i18n/blob/master/CHANGELOG.md#v2) of go-i18n.
+
+Also, it's a good idea to stay up to date with the major version of the libraries that you are using. This way you get the latest bug fixes because open libraries don't backport fixes.
+
+With all this in mind it was not a hard decision for us to go with V2 of go-i18n.
+
+
+#### Choosing a file format
+
+Go-i18n supports multiple file formats for the translation file. In fact you can use any file format as long as you implement an [unmarshal function](https://godoc.org/github.com/nicksnyder/go-i18n/v2/i18n#UnmarshalFunc) yourself. The most common formats are `JSON`, `TOML` and `YAML`.
+
+Yet again we where faced with a decision: Which format should we use? The Mattermost server uses `JSON`, but is this a reason for us to follow this decision? `goi18n` uses `TOML` by default. We didn't wanted to dive to deep into [the comparisons between these three](https://gohugohq.com/howto/toml-json-yaml-comparison/) and stuck with `JSON`. In retrospect we might have been better sicking with the default format `TOML` of `goi18n` for consistency with the library default.
+
+
+### Integrating go-i18n into Matterpoll
+
+With these decision made we started working on integration go-i18n into Matterpoll. But first we have to introduce three concepts that go-i18n uses (technical speaking they are just structs):
+
+- [`Bundle`](https://godoc.org/github.com/nicksnyder/go-i18n/v2/i18n#Bundle): A `Bundle` contains the translations
+- [`Message`](https://godoc.org/github.com/nicksnyder/go-i18n/v2/i18n#Message): A `Message` is a translation string and can contain plural rules
+- [`Localizer`](https://godoc.org/github.com/nicksnyder/go-i18n/v2/i18n#Localizer): A `Localizer` translates `Messages` to a specif language using a `Bundle`
+
+Because Matterpoll needs to fetch the translation files on startup, we need to include them into the plugin bundle. The Makefile of the plugin starter template allows plugin developers to just place there file into `assets` [to get them includes in the bundle](https://github.com/mattermost/mattermost-plugin-starter-template#how-do-i-include-assets-in-the-plugin-bundle). Since Mattermost v5.10 there is also a [`GetBundlePath()`](https://developers.mattermost.com/extend/plugins/server/reference/#API.GetBundlePath) method of the plugin API, that return the absolute path where the plugin's bundle was unpacked. This makes accessing assets much easier. The code to load the translation [looks like this](https://github.com/matterpoll/matterpoll/pull/133/files#diff-700816f9b4d51d7404d71e90d2661ddcR15-R45):
```

Comment:
```suggestion
Because Matterpoll needs to fetch the translation files on startup, we need to include them into the plugin bundle. The Makefile of the plugin starter template allows plugin developers to just place their file into `assets` [to get them included in the bundle](https://github.com/mattermost/mattermost-plugin-starter-template#how-do-i-include-assets-in-the-plugin-bundle). 

Since Mattermost v5.10 there is also a [`GetBundlePath()`](https://developers.mattermost.com/extend/plugins/server/reference/#API.GetBundlePath) method of the plugin API, that returns the absolute path where the plugin's bundle was unpacked. This makes accessing assets much easier. The code to load the translation [looks like this](https://github.com/matterpoll/matterpoll/pull/133/files#diff-700816f9b4d51d7404d71e90d2661ddcR15-R45):
```